bgpd/bgp_attr.c

```diff
@@ -2701,8 +2701,9 @@ size_t bgp_packet_mpattr_start(struct stream *s, struct peer *peer, afi_t afi,
 
 void bgp_packet_mpattr_prefix(struct stream *s, afi_t afi, safi_t safi,
 			      struct prefix *p, struct prefix_rd *prd,
-			      mpls_label_t *label, int addpath_encode,
-			      u_int32_t addpath_tx_id, struct attr *attr)
+			      mpls_label_t *label, u_int32_t num_labels,
+			      int addpath_encode, u_int32_t addpath_tx_id,
```

Comment:
you pass a mpls_label_t structure, but what it will contain may contain VNI.
VNI != MPLS label.
The encoding differs.
How do you handle that ? 

I assume you follow https://tools.ietf.org/html/draft-ietf-bess-evpn-overlay-04#5.1.3 
(in EVPN, an MPLS label is distributed by the egress PE, via the EVPN
control plane MAC Advertisement and Ethernet AD per EVI, is used to carry
the VNI. )
Right ?

Right now, using EVPN implies using VNIs.
If it is the case, then why don't you use different structure than mpls_label_t ( vni_t for instance) ?
If you still want to use label, why don't you create a union of mpls_label and vni ?

If you want to make it flexible, then you should mention in comment what should be done in order to distinguish if the so called label is a VNI or a real MPLS label ?

Reply:
I didn't really change it in this PR. The current code also has mpls_label_t. I believe the reason been EVPN can work with VxLan and MPLS. We figure out whether the label filed in EVPN carries VNI or MPLS label based on the encapsulation.

Reply:
is there a vty related to the encapsulation ?
is there a function() called in the rx data path of evpn input that checks if encapsulation is supported or not ? ( I did not see the case where the encapsulation was not supported, I would have appreciated to h ave it.)

Reply:
You make a valid point but the changes you are requesting are outside the scope of this PR. Currently, FRR only supports VxLan with EVPN and the code assumes that at multiple places. This PR doesn't change any of the existing behavior and assumptions regarding encap type. 
A config option to set the encap type and other related code would be added when we add support for MPLS or any other encap type with EVPN.

Reply:
ok. I saw that bgp_evpn_label2str() is already in the master.
it is out of the scope of this issue.
what we can do is to create an issue to track this valid point outside.
ok ?

Reply:
sure, that make sense. We should definitely create a task to track the issue.

Reply:
created a new issue in frr to track this: #1716
https://github.com/FRRouting/frr/issues/1716